### PR TITLE
Sample SAIFE ZKPR code in cpp

### PR DESCRIPTION
**Pull request by matthewwwalters**
This code should be relatively similar to the Java code that Jacob G. developed.  The goal is to provide a potential developer an idea on how the Zero Knowledge Password Reset functionality would be implemented.

### CODE REVIEW

**Comment by jabocg** [comment]
Looks good to me. +1

Sidenote, I'm piggybacking a commit on this PR. All it consists of is copyright updates and adding a README for java.

File: SAIFEPasswordReset/ZeroKnowledge/cpp/src/README.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+To build:
```
Comment (matthewwwalters) [inline SAIFEPasswordReset/ZeroKnowledge/cpp/src/README.md]: Add additional information telling the develop that the headers and libraries for the SAIFE SDK were downloaded and expanded into the current directory.  Just needs a little bit more elaboration to walk the developer through the setup and compilation process.

File: SAIFEPasswordReset/ZeroKnowledge/cpp/src/SaifeZKPR.cpp [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+ *
+ * UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, THE SOFTWARE
+ * AND DOCUMENTATION ARE DISTRIBUTED ON AN "AS IS" BASIS, WITHOUT WARRANTIES
+ * OR CONDITIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED INCLUDING BUT NOT
+ * LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
+ * PURPOSE AND NONINFRINGEMENT.  REFER TO THE WRITTEN AGREEMENT FOR SPECIFIC
+ * LANGUAGE GOVERNING PERMISSIONS AND LIMITATIONS.
+ *
+ */
+
```
Comment (matthewwwalters) [inline SAIFEPasswordReset/ZeroKnowledge/cpp/src/SaifeZKPR.cpp]: Add file header comments to describe the example, why it exists, and the outside dependencies so that the code can exist on its own.

File: SAIFEPasswordReset/ZeroKnowledge/java/README.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+# compile to bin folder
+javac -cp $CP -d bin/ src/com/saife/example/SaifeZKPR.java
+```
+
+## Running
+
+1. Run Java code
+2. Reset password via [SAIFE dashboard](https://dashboard.saifeinc.com)
+1. Password is now reset and stored in the file `saife_password`
```
Comment (joekamien) [inline SAIFEPasswordReset/ZeroKnowledge/java/README.md]: Should be 3. I know that markdown autocorrects this, but it still looks bad for anyone who views the plain text document.

File: SAIFEPasswordReset/ZeroKnowledge/cpp/src/SaifeZKPR.cpp [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+class MyPwdResetListener : public saife::PasswordResetListenerInterface
+{
+public:
+    MyPwdResetListener() { }
+    virtual ~MyPwdResetListener() { }
+    virtual void PasswordResetProcessed() {
+        std::cout << "The SDK has received and processed a password reset message." << std::endl;
+        std::string newPwd = generateNewKeystorePassword();
+        std::cout << "Changing the keystore password to the following: " << newPwd << std::endl;
+        saife_ptr->SetUserCredential(newPwd);
```
Comment (joekamien) [inline SAIFEPasswordReset/ZeroKnowledge/cpp/src/SaifeZKPR.cpp]: Consider adding a comment to reiterate just how important this line is. It's vital to set the user credential inside the callback.

File: SAIFEPasswordReset/ZeroKnowledge/cpp/src/SaifeZKPR.cpp [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+using saife::LicenseExceededException;
+using saife::NoSuchContactException;
+using saife::UnlockRequiredException;
+using saife::PresenceRequiredException;
+
+//We arbitrarily set the password to 16 random (printable) characters
+static const int keystorePasswordLength = 16;
+
+//Here's where we store the keystore password.
+static const std::string passwordFileName = "saife_password";
```
Comment (joekamien) [inline SAIFEPasswordReset/ZeroKnowledge/cpp/src/SaifeZKPR.cpp]: Why do we read/store the password in a file? Why not just prompt the user for a password and get from stdin?

File: SAIFEPasswordReset/ZeroKnowledge/cpp/src/SaifeZKPR.cpp [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+    std::string newPwd;
+
+    for (int i = 0; i < keystorePasswordLength; i++)
+        newPwd += alphanums[pick(rng)];
+
+    return newPwd;
+}
+
+//There are multiple ways of generating entropy on devices.  This code is
+//intended for illustrative purposes only.  Use at your own risk.
```
Comment (joekamien) [inline SAIFEPasswordReset/ZeroKnowledge/cpp/src/SaifeZKPR.cpp]: Do we have documentation on different ways of generating entropy? Can we provide a link?

File: SAIFEPasswordReset/ZeroKnowledge/cpp/src/SaifeZKPR.cpp [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+    }
+  } catch (InvalidManagementStateException& e) {
+    std::cerr << e.error() << std::endl;
+  } catch (SaifeInvalidCredentialException& e) {
+    std::cerr << e.error() << std::endl;
+  } catch (...) {
+    std::cerr << "Failed to initialize library with unexpected error" << std::endl;
+  }
+
+  return 1;
```
Comment (joekamien) [inline SAIFEPasswordReset/ZeroKnowledge/cpp/src/SaifeZKPR.cpp]: Should return 0 if program ran successfully.

File: SAIFEPasswordReset/ZeroKnowledge/java/README.md [hunk 8]
```diff
@@ -0,0 +1,40 @@
+# Zero Knowledge Password Reset(Java)
```
Comment (joekamien) [inline SAIFEPasswordReset/ZeroKnowledge/java/README.md]: This pull request is updating C++ code and Java readme. That's kind of confusing. In the future, can we split into two different ones?

Comment (jabocg) [inline SAIFEPasswordReset/ZeroKnowledge/java/README.md]: This was a one time thing because the Java pull request did not get the full pull request treatment. I'm going to make a new PR off of develop for Java changes, sans readme and copyright.

File: SAIFEPasswordReset/ZeroKnowledge/cpp/src/SaifeZKPR.cpp (hunk 5 above)
Comment (matthewwwalters) [inline SAIFEPasswordReset/ZeroKnowledge/cpp/src/SaifeZKPR.cpp]: Joe, we could have prompted the user for a password.  That would have been a perfectly good example as well.  I wanted to avoid the issue with the first time run (how does the user know what the password when they first run the application, or haven't run the application in a very long time and have forgotten the password).  Yes, this is a password reset example but  I guess I wanted to provide some semblance of the application "working" before we got into a locked state where the user did not know the password.

The other thing I was thinking was that the user could "tail -f saife_password" and simply watch the new passwords being generated.  Turns out that simply logging the password changes are more effective.

Long story short:  there is no good reason for not prompting the user for a password.  We simply chose to write it to a file.

File: SAIFEPasswordReset/ZeroKnowledge/cpp/src/SaifeZKPR.cpp (hunk 4 above)
Comment (matthewwwalters) [inline SAIFEPasswordReset/ZeroKnowledge/cpp/src/SaifeZKPR.cpp]: Will add a note about setting the user's credential before the keystore locks.  However, it is not required to set the user credential inside the callback.  The developer may chose to keep the keystore unlocked and set the user password in a UI thread, for example.

File: SAIFEPasswordReset/ZeroKnowledge/cpp/src/SaifeZKPR.cpp (hunk 6 above)
Comment (matthewwwalters) [inline SAIFEPasswordReset/ZeroKnowledge/cpp/src/SaifeZKPR.cpp]: Not at this time.  I opened up ticket DOCEX-8 and a systems ticket SYS-59 to document the lack of this.  For now, we will say "Contact SAIFE for further discussions".  The reason for the delay is that the information communicated to developers (the generation of randomness) could be considered technical material subject to export compliance.  The system team needs to generate documentation that can pass muster if ever challenged by an export compliance officer.

File: SAIFEPasswordReset/ZeroKnowledge/cpp/src/SaifeZKPR.cpp (hunk 7 above)
Comment (matthewwwalters) [inline SAIFEPasswordReset/ZeroKnowledge/cpp/src/SaifeZKPR.cpp]: Agreed and fixed.

File: SAIFEPasswordReset/ZeroKnowledge/cpp/src/README.md (hunk 1 above)
Comment (matthewwwalters) [inline SAIFEPasswordReset/ZeroKnowledge/cpp/src/README.md]: Completed.

File: SAIFEPasswordReset/ZeroKnowledge/cpp/src/SaifeZKPR.cpp (hunk 2 above)
Comment (matthewwwalters) [inline SAIFEPasswordReset/ZeroKnowledge/cpp/src/SaifeZKPR.cpp]: Completed.

File: SAIFEPasswordReset/ZeroKnowledge/cpp/src/SaifeZKPR.cpp (hunk 5 above)
Comment (matthewwwalters) [inline SAIFEPasswordReset/ZeroKnowledge/cpp/src/SaifeZKPR.cpp]: Completed.
